pkg/skaffold/build/cache.go

```diff
@@ -0,0 +1,373 @@
+type ArtifactCache map[string]ImageDetails
+
+// Cache holds any data necessary for accessing the cache
+type Cache struct {
+	artifactCache ArtifactCache
+	client        docker.LocalDaemon
+	cacheFile     string
+	useCache      bool
+}
+
+var (
+	// For testing
+	hashForArtifact = getHashForArtifact
+	localCluster    = config.GetLocalCluster
+	noCache         = &Cache{}
+)
+
+const (
+	prefixLength = len("sha256:")
+)
+
+// NewCache returns the current state of the cache
+func NewCache(useCache bool, cacheFile string) *Cache {
+	if !useCache {
+		return noCache
+	}
+	cf, err := resolveCacheFile(cacheFile)
+	if err != nil {
+		logrus.Warnf("Error resolving cache file, not using skaffold cache: %v", err)
+		return noCache
+	}
+	cache, err := retrieveArtifactCache(cf)
+	if err != nil {
+		logrus.Warnf("Error retrieving artifact cache, not using skaffold cache: %v", err)
+		return noCache
+	}
+	client, err := docker.NewAPIClient()
+	if err != nil {
+		logrus.Warnf("Error retrieving local daemon client, not using skaffold cache: %v", err)
+		return noCache
+	}
+	return &Cache{
+		artifactCache: cache,
+		cacheFile:     cf,
+		useCache:      useCache,
+		client:        client,
+	}
+}
+
+// resolveCacheFile makes sure that either a passed in cache file or the default cache file exists
+func resolveCacheFile(cacheFile string) (string, error) {
+	if cacheFile != "" {
+		return cacheFile, util.VerifyOrCreateFile(cacheFile)
+	}
+	home, err := homedir.Dir()
+	if err != nil {
+		return "", errors.Wrap(err, "retrieving home directory")
+	}
+	defaultFile := filepath.Join(home, constants.DefaultSkaffoldDir, constants.DefaultCacheFile)
+	return defaultFile, util.VerifyOrCreateFile(defaultFile)
+}
+
+func retrieveArtifactCache(cacheFile string) (ArtifactCache, error) {
+	cache := ArtifactCache{}
+	contents, err := ioutil.ReadFile(cacheFile)
+	if err != nil {
+		return nil, err
+	}
+	if err := yaml.Unmarshal(contents, &cache); err != nil {
+		return nil, err
+	}
+	return cache, nil
+}
+
+// RetrieveCachedArtifacts checks to see if artifacts are cached, and returns tags for cached images, otherwise a list of images to be built
+func (c *Cache) RetrieveCachedArtifacts(ctx context.Context, out io.Writer, artifacts []*latest.Artifact) ([]*latest.Artifact, []Artifact) {
+	if !c.useCache {
+		return artifacts, nil
+	}
+	color.Default.Fprintln(out, "Checking cache...")
+	var needToBuild []*latest.Artifact
+	var built []Artifact
+	for _, a := range artifacts {
+		artifact, err := c.retrieveCachedArtifact(ctx, out, a)
+		if err != nil {
+			logrus.Debugf("error retrieving cached artifact for %s: %v\n", a.ImageName, err)
+			needToBuild = append(needToBuild, a)
+			continue
+		}
+		if artifact == nil {
+			needToBuild = append(needToBuild, a)
+			continue
+		}
+		built = append(built, *artifact)
+	}
+	return needToBuild, built
+}
+
+func (c *Cache) retrieveCachedArtifact(ctx context.Context, out io.Writer, a *latest.Artifact) (*Artifact, error) {
```

Comment:
this is a complex method - and it is a bit surprising that a "retrieve*" method has a side-effect like pushing to a registry, also none of the push related logic is covered by tests. Maybe we could pull out the pull related logic to the signature as return values like "needsPush"?

Reply:
done, pulled out logic into "needsPush", "needsRebuild" and "needsRetag"